### ISSUE TITLE
This is not thread local, it is in fact fiber local.

### ISSUE THREAD

**Issue by ioquatix** [rspec/rspec-support#580]
https://github.com/rspec/rspec-support/blob/6d33eaa48a761d616e23b81fc33b68a6e68ec296/lib/rspec/support.rb#L93

I suggest introducing an actual thread local variable, e.g.

```ruby
class Thread
  attr_accessor :rspec_state
end
```

**Comment by ioquatix** [comment]
See https://github.com/socketry/async-rspec/pull/22 for an example of where this can be a problem.

**Comment by pirj** [comment]
If I understand the docs correctly, it's suggested to use:
```ruby
Thread.current.thread_variable_set(:foo, 1)
Thread.current.thread_variable_get(:foo)
```
to share the value between all fibers of the same thread. So we would have to:
```diff
- Thread.current[:__rspec] ||= {}
+ Thread.current.thread_variable_get(:__rspec) || Thread.current.thread_variable_set(:__rspec, {})
```
Does this sound right, @ioquatix ?

**Comment by ioquatix** [comment]
I think you should do the same as I proposed here: https://github.com/ruby/debug/pull/987

Such an approach was also adopted recently by Rails.

So:

```
class Thread
  attr_accessor :rspec_state
end

module RSpec::Support
  def self.current_state
    Thread.current.rspec_state ||= {}
  end
end
```

That's assuming you will always only run one test per thread at a time.

**Comment by JonRowe** [comment]
We try to avoid monkey patching as much as possible, so a solution which doesn't modify the thread class is preferred.

**Comment by ioquatix** [comment]
I wouldn't really call this a monkey patch.

**Comment by pirj** [comment]
> assuming you will always only run one test per thread at a time

I wouldn’t rely on this assumption in the long run

**Comment by ioquatix** [comment]
The alternative is to use `Fiber.storage` which is inherited across descendent threads and fibers.

**Comment by pirj** [comment]
We have to figure out something that would also work on 1.8 and 1.9
And we can add a note to replace with a modern API in RSpec 4

**Comment by JonRowe** [comment]
Fix release in `3.12.1` for any Ruby >= 2, older Rubies will maintain existing behaviour.
